### DISCUSSION TITLE
Unable to override X-Frame-Options header from config map in API Portal

### DISCUSSION THREAD

**Discussion by FranL27** [mia-platform/community#249] [Q&A] [answered]
### Which section?

Marketplace

### Which is the version?

1.16.3

### Description

Hi,
In my Dev Portal i'm trying to visualize the API Portal but since the it is in another project it is exposed under a different domain. 
This cause an `X-Frame-Options` header problem since somewhere in the API Portal this header is set to `SAMEORIGIN` which means that the FE can be served as frame only under an url of the same domain.
```
Refused to display '<URL>' in a frame because it set 'X-Frame-Options' to 'sameorigin'.
```
I've tried to override the ngnix `website.conf`with a custom config map where i've explicitly set the header to allow different domains, but the API Portal seems to ignore it (i've also checked that it was correctly mounted in the pod).
Here an example of how the config was made:
```
location / {
  include /etc/nginx/security.d/cross-site_script.conf;
  add_header 'Content-Security-Policy' "default-src 'self'; script-src 'self' 'unsafe-eval'; object-src 'none'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self'" always;
  add_header 'X-Frame-Options' "ALLOW-FROM DOMAIN-1 DOMAIN-2";

  expires $expires;

  try_files $uri $uri/index.html /index.html =404;
}
```
Do you have any ideas on how to fix that?

@theliuk fyi

**Comment by davidebianchi** [comment]
Hi @FranL27! Can you use the external (or cross project) proxy to expose the api-portal website from the same host?

**Reply by theliuk** [reply]
Hi @davidebianchi , 
I believe we can't, because we need two API portals to be available in the same dev portal: 

1. one API portal allows the user to try the APIs of the Demo environment (which are at `https://api-demo.mia-platform.eu`)
2. one API portal allows the user to try the APIs of the Preprod environment (which are at `https://api-preprod.mia-platform.eu`)

If we load the two API portals under the same domain of the Dev Portal (i.e. `https://documentation.mia-platform.eu`), both the API portals will expect the APIs to be available under `https://documentation.mia-platforme.eu`) but that's not the case. 

This setup was done specifically so that API portals were to be loaded from the same host as the APIs they allow to contact, and this setup worked for at least a year and a half. It seems to me that recent updates (I don't know, maybe of the browsers' policies) have caused this problem. 

I'd rather try to add the header `X-Frame-Options` to allow the API Portal exposed at `https://api-{env}.mia-platform.eu` to be loaded under `https://documentation.mia-platform.eu`).

Thanks,

Luca

**Reply by theliuk** [reply]
It's the API gateway that, for whatever reason, overrides out the X-Frame-Options header.
We did this experiment:

In a local directory we created this file `website.conf` with the following content:

```nginx
location / {
  include /etc/nginx/security.d/cross-site_script.conf;
  add_header "X-Frame-Options" "ALLOW-FROM https://google.com";
  add_header 'Content-Security-Policy' "default-src 'self'; script-src 'self' 'unsafe-eval'; object-src 'none'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; font-src 'self'" always;

  expires $expires;

  try_files $uri $uri/index.html /index.html =404;
}
```

Then we run an API portal instance locally:

`docker run -it --rm -p 8080:8080 --name web -v ~/nginx:/etc/nginx/conf.d nexus.mia-platform.eu/api-portal/website:1.16.3 `

And the trick actually works (obviously), because we get the header `X-Frame-Options` in the response.

**Reply by theliuk** [reply]
Is there a way we can tell the API gateway not to override the header, or at least to add it the way we want?

**Reply by davidebianchi** [reply]
In the api-gateway, we set some security header by default which also include the X-Frame-Options set to 'SAMEORIGIN'. You should override it also in the location of the api-gateway, using the advanced extension in Console for example.

**Reply by theliuk** [reply]
Thank you @davidebianchi. Could you point us the right place of the advanced config where to put this override? 

Thank you.

**Reply by davidebianchi** [reply, marked as answer]
You should use the `root-location-extension.conf` extension if you need to remove x-frame-options for all the project.
If you need to disable only for the website location, in the extension `server-extension.conf` you should do something like:

```nginx
location /path/to/your/fe {
  include /etc/nginx/customization.d/header-debug.conf;
  include /etc/nginx/customization.d/auth-set-userid.conf;
  include /etc/nginx/set_proxy_headers.conf;
  include /etc/nginx/security.d/*.conf;
  include /etc/nginx/customization.d/root-location-extension.conf;
  add_header "X-Frame-Options" "ALLOW-FROM ...";

  proxy_pass http://$proxy_name$proxy_url;
}
```
